In [151]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [152]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

data = pd.read_csv("science_data_large.csv")

print(data.head(15))
data.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [153]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

X = data[['Temperature °C', 'Mols KCL']]
y = data['Size nm^3']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=1234)

## Part 3. Perform a Linear Regression

In [154]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
prediction = model.predict(X_test.iloc[[1]])
print(f"Prediction: {prediction}")

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print(f"Score: {score}")

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficent = model.coef_
print(f"Coefficent: {coefficent}")

intercept = model.intercept_
print(f"Intercept: {intercept}")

Prediction: [373376.62139023]
Score: 0.8289425946295386
Coefficent: [ 877.88026456 1031.63477913]
Intercept: -414439.15050166176


The model's score of approximately 82.89% represents its accuracy on the test dataset. This means that the model's predictions matched the actual outcomes 82.89% of the time, reflecting how well it performed on unseen data.

An accuracy score of 82.89% is generally strong in many real-world applications, indicating the model's ability to generalize from the training data to new cases. For every 100 predictions made, about 83 are expected to be correct.

$h(x) = -414439 + 877.88x_1 + 1031.6x_2$

Where, $x_1$ is Temperature and $x_2$ is Mols

## Part 4. Use Cross Validation

In [155]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
cross_val_score = cross_val_score(model, X_train, y_train, cv=20)
print(cross_val_score)
print(f"Mean of cross validation score: {cross_val_score.mean()}")

# Report on their finding and their significance

[0.85719752 0.87449809 0.8363995  0.89581253 0.79773227 0.86912548
 0.81410604 0.77695811 0.85696791 0.86757316 0.78704576 0.82912954
 0.84323198 0.86707108 0.90594077 0.87725957 0.89890812 0.86908916
 0.8932177  0.8680652 ]
Mean of cross validation score: 0.8542664742596611


The model's previous accuracy score of 82.89% was based on a single train-test split, while the mean accuracy from the cross-validation is slightly higher at 85.43%. This comparison suggests that the model's performance improves when evaluated over multiple data shuffles, providing a more reliable estimate of its ability to generalize to new data.

## Part 5. Using Polynomial Regression

In [156]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

model2 = LinearRegression()
model2.fit(X_train_poly, y_train)

prediction2 = model2.predict(X_test_poly)

score_poly = model2.score(X_test_poly, y_test)
print(f"Score: {score_poly}")

print(f"coefficents: {model2.coef_}")
print(f"intercept: {model2.intercept_}")

# Report on the metrics and output the resultant equation as you did in Part 3.

Score: 1.0
coefficents: [ 0.00000000e+00  1.20000000e+01 -1.17440629e-07  6.20014255e-13
  2.00000000e+00  2.85714287e-02]
intercept: 1.707382034510374e-05


The polynomial regression model of degree 2 yielded a perfect score of 1.0, indicating that it fit the test data perfectly. While this result seems ideal, it raises concerns about potential overfitting, as such high accuracy is uncommon in practical scenarios. 

$f(x) = 1.0 + 12x_1 + 2x_1x_2 + 0.02857x_2^2$

Where, $x_1$ is Temperature and $x_2$ is Mols